### PR TITLE
KAFKA-2598: Adding integration test for the authorizer at API level. …

### PR DESCRIPTION
**Pull request by Parth-Brahmbhatt**
…Some bug fixes that I encountered while running the tests.

### CODE REVIEW

**Comment by Parth-Brahmbhatt** [comment]
@junrao I have attached the API level tests. Sorry for the delay. 

There are 3 things that are still pending because of which you will see some TODOs and commented code in the PR:

1) Both producer and consumer side seem to hide server side exception when they fail to get Metadata. Consumer is kind of worst as it tries infinitely. Really no way to test that we actually got authorization exception or not. Shouldn't they stop retrying for nonRetriable exception types?

2) For APIs where we just throw an exception instead of sending a response back, the client does not really get any response. In the original PR I think we made the change under the assumption that when the API layer throws the exception the RequestHandler will somehow translate that into a response but that is not happening. I think it is better to just send a response back to the client.

3) This is unrelated to PR TODOs but I was looking at the https://cwiki.apache.org/confluence/display/KAFKA/KIP-11+-+Authorization+Interface#KIP-11-AuthorizationInterface-Operation and as seen in the table operations like LeaderAndIsr, StopReplica and UpdateMetadata were defined as ClusterAction at topic level. In the current code all of these are defined as ClusterAction at Cluster level. I don't know if this was changed as part of KIP discussion, PR comments or its just a complete miss on my end. I wanted to point this out and ensure that we want these operations to be authorized at cluster level with ClusterAction operation and not do what is listed in the KIP.

**Comment by ijuma** [comment]
@Parth-Brahmbhatt, regarding 1 there is a JIRA issue to track it (https://issues.apache.org/jira/browse/KAFKA-2391), which needs to be resolved before 0.9.0.0. My suggestion is to file a separate JIRA for the tests that depend on KAFKA-2391 with a target of 0.9.0.0 (and with a reference to KAFKA-2391).

In this PR, we can fix the issues that have been found and the tests that don't depend on KAFKA-2391.

**Comment by junrao** [comment]
1. Yes, we can file separate jiras to address those.
2. Hmm, in KafkaApi.handle(), we have a general try/catch block to send an error response when we get an exception. Could you trace into it to see if that logic gets called?
3. I think for ClusterAction, the only resource should be Cluster. Could you update the KIP wiki?

**Comment by Parth-Brahmbhatt** [comment]
Addressed 2 and 3.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -490,6 +508,9 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
                     else
                         this.offsetOutOfRangePartitions.put(tp, fetchOffset);
                     log.info("Fetch offset {} is out of range, resetting offset", subscriptions.fetched(tp));
+                } else if (partition.errorCode == Errors.AUTHORIZATION_FAILED.code()) {
+                    log.info("not authorized to read from topic {}.", tp.topic());
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Minor, but our logging messages start with a capital letter.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Also, is it correct that this is logged at `info` level?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    incoming.readFully(response)
+    ByteBuffer.wrap(response)
+  }
+
+  private def sendRecords(numRecords: Int, tp: TopicPartition) {
+    val futures = (0 until numRecords).map { i =>
+      this.producers.head.send(new ProducerRecord(tp.topic(), tp.partition(), i.toString.getBytes, i.toString.getBytes))
+    }
+    try {
+      futures.map(_.get)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: As Jun said, `foreach` is better for cases where we don't use the return value.

File: core/src/main/scala/kafka/common/AuthorizationException.scala [hunk 3]
```diff
@@ -21,4 +21,5 @@ package kafka.common
  * @param message
  */
 class AuthorizationException(message: String) extends RuntimeException(message) {
+  def this() = this(null)
```
Comment (ijuma) [inline core/src/main/scala/kafka/common/AuthorizationException.scala:24]: Why is this change needed?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -288,6 +291,20 @@ private void throwIfOffsetOutOfRange() throws OffsetOutOfRangeException {
     }
 
     /**
+     * If any topic from previous fetchResponse contatains Authorization error, throw ApiException.
+     * @throws ApiException
+     */
+    private void throwIfUnauthorized() throws ApiException {
+        if (!unauthorizedTopicPartition.isEmpty()) {
+            StringBuilder sb = new StringBuilder();
+            for (TopicPartition topicPartition: unauthorizedTopicPartition)
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Minor: space before colon is missing (I know, the fact that this is different for the Java versus Scala code is annoying).

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 5]
```diff
@@ -80,6 +81,7 @@
     private final Deserializer<V> valueDeserializer;
 
     private final Map<TopicPartition, Long> offsetOutOfRangePartitions;
+    private final Set<TopicPartition> unauthorizedTopicPartition;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Should be `unauthorizedTopicPartitions` right?

File: core/src/main/scala/kafka/api/ControlledShutdownRequest.scala [hunk 6]
```diff
@@ -58,7 +58,7 @@ case class ControlledShutdownRequest(versionId: Short,
   def sizeInBytes: Int = {
     2 + /* version id */
       4 + /* correlation id */
-      clientId.fold(0)(shortStringLength)
+      clientId.fold(0)(shortStringLength) +
```
Comment (ijuma) [inline core/src/main/scala/kafka/api/ControlledShutdownRequest.scala:61]: Ouch, nice catch. I'm surprised that the tests that trigger controlled shutdown did not fail.

Comment (ijuma) [inline core/src/main/scala/kafka/api/ControlledShutdownRequest.scala:61]: Oh, I understand. It's because we use the Java implementation of `ControlledShutdownRequest` unless "inter.broker.protocol.version" is < 0.9.0.0. So, we would only have caught this via the (in-progress) upgrade tests.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 7]
```diff
@@ -593,7 +593,7 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     val unauthorizedTopicMetaData = unauthorizedTopics.map(topic => new TopicMetadata(topic, Seq.empty[PartitionMetadata], ErrorMapping.AuthorizationCode))
 
-    val topicMetadata = getTopicMetadata(authorizedTopics, request.securityProtocol)
+    val topicMetadata = if (authorizedTopics.isEmpty) Seq.empty[TopicMetadata] else getTopicMetadata(authorizedTopics, request.securityProtocol)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:609]: Probably out of scope for this PR, but we should really change `getTopicMetadata`'s contract not to assume that empty topics means all topics as it's quite error-prone.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+        consumeRecords(this.consumers.head)
+        Assert.fail("should have thrown exception")
+      } catch {
+        case e: ApiException => Assert.assertEquals("Not authorized to read from topic-0", e.getMessage)
+      }
+    }
+
+    @Test
+    def testAllowingReadOnTopicAndGroupAllowsReading() {
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:271]: Would there be a benefit to using a KafkaPrincipal from a SSL certificate as suggested by Jun?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 1 above)
Comment (Parth-Brahmbhatt) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Fixed, changed the log level to warn.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 2 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Done.

File: core/src/main/scala/kafka/common/AuthorizationException.scala (hunk 3 above)
Comment (Parth-Brahmbhatt) [inline core/src/main/scala/kafka/common/AuthorizationException.scala:24]: Because of this line https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/common/ErrorMapping.scala#L96 that uses reflection to create an instance of the Exception class.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 4 above)
Comment (Parth-Brahmbhatt) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Fixed. I thought in java our checkstyle would report things like this, guess we need to update our template.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 5 above)
Comment (Parth-Brahmbhatt) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Fixed.

File: core/src/main/scala/kafka/api/ControlledShutdownRequest.scala (hunk 6 above)
Comment (Parth-Brahmbhatt) [inline core/src/main/scala/kafka/api/ControlledShutdownRequest.scala:61]: Yep, took me sometime to figure out why I keep getting BufferOverflow.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 7 above)
Comment (Parth-Brahmbhatt) [inline core/src/main/scala/kafka/server/KafkaApis.scala:609]: +1,  Jun pointed this issue out to me in the initial PR and I thought I handled it correctly, Thank god for test cases but I feel like a complete idiot for missing this even after being specifically told :-(.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 8 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:271]: I originally thought I need to have greater than 1 identity but as I wrote tests I did not see the need. Using SSL certificate will probably test the SSL auth part in addition to authorizer but will not change the validity of these tests in anyway. Let me know if you think otherwise.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 9]
```diff
@@ -288,6 +291,20 @@ private void throwIfOffsetOutOfRange() throws OffsetOutOfRangeException {
...
+     * If any topic from previous fetchResponse contatains Authorization error, throw ApiException.
+     * @throws ApiException
+     */
+    private void throwIfUnauthorized() throws ApiException {
+        if (!unauthorizedTopicPartitions.isEmpty()) {
+            StringBuilder sb = new StringBuilder();
+            for (TopicPartition topicPartition : unauthorizedTopicPartitions)
+                sb.append(topicPartition + ",");
+            unauthorizedTopicPartitions.clear();
+            throw new ApiException(String.format("Not authorized to read from %s", sb.substring(0, sb.length() - 1).toString()));
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Perhaps it's better to have an explicit AuthorizationException and just throw that.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+      for ((resource, acls) <- RequestKeysToAcls(key))
+        addAndVerifyAcls(acls, resource)
+
+      sendRequestAndVerifyResponseErrorCode(socket, key, requestBytes, ErrorMapping.NoError)
+    }
+  }
+
+    @Test
+    def testProduceNeedsAuthorization() {
+    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Describe)), topicResource)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: indentation

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+        sendRecords(numRecords, tp)
+        Assert.fail("should have thrown exception")
+      } catch {
+        case e: ApiException => Assert.assertEquals(Errors.AUTHORIZATION_FAILED.exception().getMessage, e.getMessage)
+      }
+    }
+
+    @Test
+    def testOnlyWritePermissionAllowsWritingToProducer() {
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:230]: Hmm, don't we need the ACL on describe topic to be able to produce?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+    def testCreatePermissionNeededForWritingToNonExistentTopic() {
+      val newTopic = "newTopic"
+      val topicPartition = new TopicPartition(newTopic, 0)
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), new Resource(Topic, newTopic))
+      try {
+        sendRecords(numRecords, topicPartition)
+        Assert.fail("should have thrown exception")
+      } catch {
+        case e: TimeoutException =>
+        //TODO Need to update the producer so it actually throws the server side of exception.
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:244]: Could we file a separate jira to track this? Also, the default request timeout is 30secs. This is probably too long for the unit test, we should reduce that.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+        sendRecords(numRecords, topicPartition)
+      } catch {
+        case e: Exception => Assert.fail("No exception should be thrown.")
+      }
+    }
+
+    @Test
+    def testConsumerNeedsAuthorization() {
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
+      //TODO: Ideally we would want to test that when consumerGroup permission is not present we still get an AuthorizationException
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:256]: Could we file a separate jira for that?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 8 above)
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:271]: Could we add at least one end to end test to make sure we can do ACL based on SSL peer principal?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+//        consumeRecords(this.consumers(0))
+//      } catch {
+//        case e: Exception => Assert.fail("No exception should be thrown.")
+//      }
+//    }
+
+  def removeAllAcls() = {
+    authorizer.getAcls().keys.foreach { resource =>
+      authorizer.removeAcls(resource)
+      TestUtils.waitAndVerifyAcls(Set.empty[Acl], authorizer, resource)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: To really makes sure that the acl is propagated to the broker, it seems that we need to verify against the authorizer instance in the broker.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+    try {
+      futures.foreach(_.get)
+    } catch {
+      case e: ExecutionException => throw e.getCause
+    }
+  }
+
+  private def addAndVerifyAcls(acls: Set[Acl], resource: Resource) = {
+    authorizer.addAcls(acls, resource)
+    TestUtils.waitAndVerifyAcls(authorizer.getAcls(resource) ++ acls, authorizer, resource)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: To really makes sure that the acl is propagated to the broker, we need to verify against the authorizer instance in the broker.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+        case e: ApiException => Assert.assertEquals("Not authorized to read from topic-0", e.getMessage)
+      }
+    }
+
+    @Test
+    def testAllowingReadOnTopicAndGroupAllowsReading() {
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), topicResource)
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), groupResource)
+      sendRecords(1, tp)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:274]: Don't we need an ACL on describe topic to consume?

**Comment by ijuma** [comment]
@Parth-Brahmbhatt, are you still working on this? Also, there are some conflicts in this branch that need to be resolved.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+      }
+    }
+
+    @Test
+    def testOnlyWritePermissionAllowsWritingToProducer() {
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
+      try {
+        sendRecords(numRecords, tp)
+      } catch {
+        case e: Exception => Assert.fail("Should not fail")
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Isn't it better to not catch the exception at all? It seems like it would have more useful information than the `Assert.fail` and it's less code.

Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: There are a few other instances of this. I think it's fine to catch and rethrow if we add some useful information. Otherwise, it doesn't seem worth it.

**Comment by junrao** [comment]
@Parth-Brahmbhatt : The SASL patch is now committed. Ideally, we want to be able to test ACL based on both SSL and SASL principals. However, if it's too much in the jira, perhaps we can at least test the ACL with SASL principals, if that's easier. Does that sound good to you?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 9 above)
Comment (Parth-Brahmbhatt) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Done.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 10 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Fixed.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 11 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:230]: Read and Write gives Describe by default.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 12 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:244]: I am not seeing 30 seconds, the test finishes pretty quickly.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 13 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:256]: doesen't https://issues.apache.org/jira/browse/KAFKA-2391 cover this?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 14 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Done.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 15 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Done

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 16 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:274]: Read access covers Describe.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 17 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Fixed.

**Comment by Parth-Brahmbhatt** [comment]
@junrao I am fine with adding sasl + ssl test for end to end coverage, not sure if we want to do that in this single jira. There is a setup overhead for both 2 way ssl and Sasl with MiniKDC. I can refactor this test so it will be a super class, we can establish server identities in the extension of this test class and the base class will have all the tests that will execute for each type of identity. 

That being said, I am not sure if this test is really the right place to do any of that. How we establish identity is irrelevant to the authorizer because as long as authorizer gets some identity we have enough testing around it to ensure it will do the authorization correctly. With the recommended tests, we are just going to test that right identities will be established which in my opinion should be tested as part of sasl or ssl testing.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 13 above)
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:256]: That JIRA is just about making sure that the consumer doesn't block. It is likely to throw a TimeoutException once it's fixed. We should have a JIRA that depends on KAFKA-2391 for the authorization part.

**Comment by ijuma** [comment]
I agree with Jun that an end to end test would be useful. In my opinion, it doesn't have to test all the endpoints, but it should check that the different pieces work well together. Having said that, I think it could be done in a separate JIRA. It would be really good to get the fixes in this PR merged.

File: clients/src/main/java/org/apache/kafka/common/errors/AuthorizationException.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+import org.apache.kafka.common.KafkaException;
+
+public class AuthorizationException extends KafkaException {
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/common/errors/AuthorizationException.java]: It seems that AuthorizationException should extend ApiException.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+      new OffsetAndMetadata(new OffsetMetadata(0)))),
+    RequestKeys.OffsetFetchKey -> new OffsetFetchRequest(groupId = group, requestInfo = Seq(topicAndPartition)),
+    RequestKeys.GroupMetadataKey -> new GroupMetadataRequest(group = group),
+    RequestKeys.UpdateMetadataKey -> UpdateMetadataRequest(UpdateMetadataRequest.CurrentVersion, correlationId, clientId, brokerId,
+      Int.MaxValue, Map(topicAndPartition -> partitionStateInfo), Set(new Broker(brokerId, Map(SecurityProtocol.PLAINTEXT -> endPoint)))),
+    RequestKeys.LeaderAndIsrKey -> new LeaderAndIsrRequest(LeaderAndIsrRequest.CurrentVersion, correlationId, clientId,
+      brokerId, Int.MaxValue, Map((topic, part) -> partitionStateInfo), Set(new BrokerEndPoint(brokerId,"localhost", 0))),
+    RequestKeys.StopReplicaKey -> StopReplicaRequest(StopReplicaRequest.CurrentVersion, correlationId, clientId, brokerId,
+      Int.MaxValue, true, Set(topicAndPartition)),
+    RequestKeys.ControlledShutdownKey -> ControlledShutdownRequest(ControlledShutdownRequest.CurrentVersion, correlationId, Some(clientId), brokerId)
```
Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Could we cover the following requests too?

  val JoinGroupKey: Short = 11
  val HeartbeatKey: Short = 12
  val LeaveGroupKey: Short = 13
  val SyncGroupKey: Short = 14

**Comment by junrao** [comment]
I agree that in unit tests, it's enough to just verify that we can get the expected user name on the broker side through either SSL or SASL. This can be done in a separate jira. Just left a couple of more comments above.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 12 above)
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:244]: I filed https://issues.apache.org/jira/browse/KAFKA-2691 which is for both the wrong exception in the producer and the hanging in the consumer. Once KAFKA-2391 is done, the issue becomes the same for both consumer and producer.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 13 above)
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:256]: I filed https://issues.apache.org/jira/browse/KAFKA-2691 which is for both the wrong exception in the producer and the hanging in the consumer. Once KAFKA-2391 is done, the issue becomes the same for both consumer and producer.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+    def testAllowingReadOnTopicAndGroupAllowsReading() {
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), topicResource)
+      addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Read)), groupResource)
+      sendRecords(1, tp)
+      this.consumers.head.assign(List(tp).asJava)
+      consumeRecords(this.consumers.head)
+    }
+  
+//    TODO: The following test goes into an infinite loop as consumer waits for consumer metadata to be propogated for ever.
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:279]: @Parth-Brahmbhatt I think the problem is that AbstractCoordinator.ensureCoordinatorKnown is not propagating non-retriable exceptions. We should be able to fix this relatively easily with something like this:

``` java
    public void ensureCoordinatorKnown() {
        while (coordinatorUnknown()) {
            RequestFuture<Void> future = sendGroupMetadataRequest();
            client.poll(future, requestTimeoutMs);

            if (future.failed() && !future.isRetriable())
                throw future.exception();

            client.awaitMetadataUpdate();
        }
    }
```

Want to give that a try and see if it fixes this issue?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 19 above)
Comment (Parth-Brahmbhatt) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Tried to add these and I am pretty confused why we have client side request and response (which are in java) and server side request and response (in scala) and even in that the 4 APIS defined above are unique as they don't have the scala counter part so I have to either give up the type checking and do something special for these 4 requests/responses that do not extend ReqeustOrResponse. 

Do we plan to get rid of the scala request responses at some point given we now have a common network layer that both core and client should be using?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 21]
```diff
@@ -303,6 +302,19 @@ private void throwIfOffsetOutOfRange() throws OffsetOutOfRangeException {
     }
 
     /**
+     * If any topic from previous fetchResponse contatains Authorization error, throw ApiException.
+     * @throws ApiException
+     */
+    private void throwIfUnauthorized() throws ApiException {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:308]: Just to check my understanding, we wouldn't actually expect to hit this case unless the user has Describe access, but not Read access on any of its subscribed topics. If the user doesn't have Describe access, then we wouldn't be able to fetch metadata for the topic and we therefore wouldn't do any fetches. In that case, the user generally wouldn't see any errors because NetworkClient just logs any errors in metadata fetches. Is that expected behavior?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 19 above)
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Yes, the Scala classes will go away: https://issues.apache.org/jira/browse/KAFKA-1927

Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: I think you may want to do something special for these 4 (which will become the approach we use for all requests in the future).

Comment (junrao) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: @Parth-Brahmbhatt , yes, the scala requests/responses will go away in the future. So, it would be better if you could rewrite the test using the java requests/responses. Sorry for not pointing this out earlier.

**Comment by junrao** [comment]
@Parth-Brahmbhatt : @hachikuji mentioned this. In https://cwiki.apache.org/confluence/display/KAFKA/KIP-11+-+Authorization+Interface, it says that we need describe on topic in order to issue GroupMetadataRequest. However, in the code, we check read permission on group. Neither seems accurate. It seems that we need to check describe permission on group?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 22]
```diff
@@ -334,6 +346,7 @@ private void throwIfRecordTooLarge() throws OffsetOutOfRangeException {
         } else {
             Map<TopicPartition, List<ConsumerRecord<K, V>>> drained = new HashMap<>();
             throwIfOffsetOutOfRange();
+            throwIfUnauthorized();
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:349]: Can you update the javadoc on KafkaConsumer.poll() to include this exception?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 20 above)
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:279]: This fix was needed in #366 also. If that gets in first, then you don't need to change it here.

**Comment by junrao** [comment]
@Parth-Brahmbhatt : We are getting pretty close to cutting the 0.9.0 release branch. Would you be able to submit a new patch based on the review feedback?

**Comment by Parth-Brahmbhatt** [comment]
@junrao Sorry for the delay , will send a PR today.

**Comment by Parth-Brahmbhatt** [comment]
@junrao Yes originally I proposed to have Describe for Group but I think Jay suggested in one of KIP meetings to just go with Read for groups. Don't remember if it was Jay or someone else. If you think we need to have Describe for Group I can change it back to that, I leave the decision up to you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 22 above)
Comment (Parth-Brahmbhatt) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:349]: Done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 21 above)
Comment (Parth-Brahmbhatt) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:308]: The correct behavior is that if you don't have Describe or Read access you should get back an AuthorizationException. Currently I handled the Read part but the consumer metadata waiting forever issue is still present and I believe we have a followup jira to fix it. We need to ensure that the metadata receiver throws the exception back to the caller when a nonRetriableException is thrown.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:308]: To clarify, you mean topic metadata, right? I don't think consumer metadata is an issue here.

**Comment by junrao** [comment]
Thanks for the patch. LGTM. We can leave GroupMetadataRequest with READ for group as it is now. We can revisit it afer KIP-40. Could you update KIP-11's wiki?
